src/test/java/io/netty/incubator/codec/http3/Http3FrameToHttpObjectCodecTest.java

```diff
@@ -652,6 +659,140 @@ public void testEncodeTrailersCompletes() {
+            } else {
+                if (hasTrailers || nonEmptyContent) {
+                    // not supported by the netty HTTP/1 model
+                    return;
+                }
+                msg = new DefaultHttpRequest(
+                        HttpVersion.HTTP_1_1, HttpMethod.POST, "/foo", new DefaultHttpHeaders());
+            }
+        } else {
+            if (last) {
+                msg = new DefaultLastHttpContent(content, trailers);
+            } else {
+                if (hasTrailers) {
+                    // makes no sense
+                    return;
+                }
+                msg = new DefaultHttpContent(content);
+            }
+        }
+
+        List<ChannelPromise> framePromises = new ArrayList<>();
+        EmbeddedQuicStreamChannel ch = new EmbeddedQuicStreamChannel(
+                new ChannelOutboundHandlerAdapter() {
+                    @Override
+                    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                        framePromises.add(promise);
+                        ctx.write(msg, ctx.voidPromise());
+                    }
+                },
+                new Http3FrameToHttpObjectCodec(false)
+        );
+
+        ChannelFuture fullPromise = ch.writeOneOutbound(msg, voidPromise ? ch.voidPromise() : ch.newPromise());
+        ch.flushOutbound();
+
+        if (headers) {
+            Http3HeadersFrame headersFrame = ch.readOutbound();
+            assertThat(headersFrame.headers().scheme().toString(), is("https"));
+            assertThat(headersFrame.headers().method().toString(), is("POST"));
+            assertThat(headersFrame.headers().path().toString(), is("/foo"));
+        }
+        if (nonEmptyContent) {
+            Http3DataFrame dataFrame = ch.readOutbound();
+            assertThat(dataFrame.content().readableBytes(), is(1));
+        }
+        if (hasTrailers) {
+            Http3HeadersFrame trailersFrame = ch.readOutbound();
+            assertThat(trailersFrame.headers().get("foo"), is("bar"));
+        }
+        // empty LastHttpContent has no data written and will complete the promise immediately
+        boolean anyData = hasTrailers || nonEmptyContent || headers || !last;
+        if (!voidPromise) {
+            if (anyData) {
+                assertFalse(fullPromise.isDone());
+            } else {
+                // nothing to write, immediately complete
+                assertTrue(fullPromise.isDone());
+            }
+        }
+        if (!last || anyData) {
+            assertFalse(ch.isOutputShutdown());
+        }
+        for (ChannelPromise framePromise : framePromises) {
+            framePromise.trySuccess();
+        }
+        if (last) {
+            assertTrue(ch.isOutputShutdown());
+        }
+        if (!voidPromise) {
+            assertTrue(fullPromise.isDone());
+        }
+    }
```

Comment:
```suggestion
            assertFalse(ch.finish());
    }
```

Reply:
hm i think this should be true, let me check

Reply:
nvm